### PR TITLE
Only send rooms with updates down sliding sync

### PR DESCRIPTION
**Pull request by erikjohnston**
Rather than always including all rooms in range.

Also adds a pre-filter to rooms that checks the stream change cache to see if anything  might have happened.

Based on #17447

### CODE REVIEW

File: synapse/handlers/sliding_sync.py [hunk 1]
```diff
@@ -617,6 +617,31 @@ async def current_sync_for_user(
...
+                    from_token.connection_position,
+                    room_id,
+                )
+                if status.status != HaveSentRoomFlag.LIVE:
+                    rooms_should_send.add(room_id)
+
+            # We only need to check for any new events and not state changes, as
+            # state changes can only happen if an event has also been sent.
+            rooms_that_have_updates = (
+                self.store._events_stream_cache.get_entities_changed(
```
Comment (erikjohnston) [inline synapse/handlers/sliding_sync.py:636]: Oh, I should move this into the DataStore rather than using the private attribute, but I don't think that should block review.

File: synapse/handlers/sliding_sync.py [hunk 2]
```diff
@@ -617,6 +617,31 @@ async def current_sync_for_user(
         # Fetch room data
         rooms: Dict[str, SlidingSyncResult.RoomResult] = {}
 
+        # Filter out rooms that haven't received updates and we've sent down
+        # previously.
```
Comment (MadLittleMods) [inline synapse/handlers/sliding_sync.py:621]: Reading the Sliding Sync [MSC3575](https://github.com/matrix-org/matrix-spec-proposals/pull/3575), I don't see this as allowed. In the spec, the `delta_tokens` only applied to `timeline`/`required_state`.

I'm assuming this is another "Simplified" Sliding Sync change we're making.

File: tests/rest/client/test_sync.py [hunk 3]
```diff
@@ -3540,19 +3540,7 @@ def test_rooms_ban_incremental_sync2(self) -> None:
...
-            channel.json_body["rooms"][room_id1]["num_live"],
-            0,
-            channel.json_body["rooms"][room_id1],
-        )
-        # There aren't anymore events to paginate to in this range
-        self.assertEqual(
-            channel.json_body["rooms"][room_id1]["limited"],
-            False,
-            channel.json_body["rooms"][room_id1],
-        )
```
Comment (MadLittleMods) [inline tests/rest/client/test_sync.py:3555]: *Tests be changing* 🫠

Context: https://github.com/element-hq/synapse/pull/17452#discussion_r1681362856

File: tests/rest/client/test_sync.py [hunk 4]
```diff
@@ -4745,6 +4736,50 @@ def test_incremental_sync_full_state_new_room(self) -> None:
...
+            self.sync_endpoint + f"?pos={after_room_token}",
+            content=sync_body,
+            access_token=user1_tok,
+        )
+        self.assertEqual(channel.code, 200, channel.json_body)
+
+        # Nothing has happened in the room, so the room should not come down
+        # /sync.
+        self.assertIsNone(channel.json_body["rooms"].get(room_id1))
+
```
Comment (MadLittleMods) [inline tests/rest/client/test_sync.py:4783]: Seems like the `test_wait_for_new_data_timeout` that currently has a `@skip_unless` might be able to be enabled now

File: tests/rest/client/test_sync.py [hunk 5]
```diff
@@ -4745,6 +4736,50 @@ def test_incremental_sync_full_state_new_room(self) -> None:
...
+        sync_body = {
+            "lists": {
+                "foo-list": {
+                    "ranges": [[0, 1]],
+                    "required_state": [
+                        [EventTypes.Create, ""],
+                        [EventTypes.RoomHistoryVisibility, ""],
+                        # This one doesn't exist in the room
+                        [EventTypes.Tombstone, ""],
+                    ],
```
Comment (MadLittleMods) [inline tests/rest/client/test_sync.py:4762]: We're not testing `required_state` so we don't need to request any

```suggestion
                    "required_state": [],
```

File: tests/rest/client/test_sync.py [hunk 6]
```diff
@@ -4745,6 +4736,50 @@ def test_incremental_sync_full_state_new_room(self) -> None:
...
+
+        _, after_room_token = self.do_sync(sync_body, tok=user1_tok)
+
+        # Make the Sliding Sync request
+        channel = self.make_request(
+            "POST",
+            self.sync_endpoint + f"?pos={after_room_token}",
+            content=sync_body,
+            access_token=user1_tok,
+        )
```
Comment (MadLittleMods) [inline tests/rest/client/test_sync.py:4776]: Can start using the `do_sync()` helper on new tests like this

File: synapse/handlers/sliding_sync.py [hunk 7]
```diff
@@ -631,7 +656,9 @@ async def handle_room(room_id: str) -> None:
                 to_token=to_token,
             )
 
-            rooms[room_id] = room_sync_result
+            # Filter out empty room results.
+            if room_sync_result:
+                rooms[room_id] = room_sync_result
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/handlers/sliding_sync.py:661]: That's true, though for initial syncs they will always be non-empty (as `initial` will be set to true). Though we should add a `and from_token` to it anyway.

Can add a test to ensure it comes down even if timeline_limit is 0 and required_state is empty.

File: synapse/handlers/sliding_sync.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/handlers/sliding_sync.py:621]: From discussion elsewhere: the proxy won't send down rooms where there haven't been changes. I think the confusion is that the list ops that the proxy will use in that case is not `sync` but a `delete / insert`.

Since we're getting rid of list ops then this distinction doesn't matter, and its ok for us to filter stuff out.

File: tests/rest/client/test_sync.py (hunk 3 above)
Comment (erikjohnston) [reply tests/rest/client/test_sync.py:3555]: The semantics be changing :grinning:

File: tests/rest/client/test_sync.py (hunk 4 above)
Comment (erikjohnston) [reply tests/rest/client/test_sync.py]: Done! Though did require changing `SlidingSyncResult.__bool__` to ignore `lists`

File: synapse/handlers/sliding_sync.py [hunk 8]
```diff
@@ -619,6 +619,31 @@ async def current_sync_for_user(
...
+                status = await self.connection_store.have_sent_room(
+                    sync_config,
+                    from_token.connection_position,
+                    room_id,
+                )
+                if status.status != HaveSentRoomFlag.LIVE:
+                    rooms_should_send.add(room_id)
+
+            # We only need to check for any new events and not state changes, as
+            # state changes can only happen if an event has also been sent.
```
Comment (MadLittleMods) [inline synapse/handlers/sliding_sync.py:636]:
```suggestion
            # We only need to check for new events since any state changes,
            # will also come down as new events.
```

File: synapse/handlers/sliding_sync.py [hunk 9]
```diff
@@ -619,6 +619,31 @@ async def current_sync_for_user(
...
+            rooms_that_have_updates = self.store.get_rooms_that_might_have_updates(
+                relevant_room_map.keys(), from_token.stream_token.room_key
```
Comment (MadLittleMods) [inline synapse/handlers/sliding_sync.py:638]: We could make the optimization here to only check rooms not already added to `rooms_should_send`:

```suggestion
                set(relevant_room_map.keys()).difference(rooms_should_send), from_token.stream_token.room_key
```

It looks like `get_rooms_that_might_have_updates(...)` is already checking an in-memory cache so maybe it's not worth it.

File: synapse/handlers/sliding_sync.py [hunk 10]
```diff
@@ -619,6 +619,31 @@ async def current_sync_for_user(
...
+        if from_token:
+            rooms_should_send = set()
+            for room_id in relevant_room_map:
+                status = await self.connection_store.have_sent_room(
+                    sync_config,
+                    from_token.connection_position,
+                    room_id,
+                )
+                if status.status != HaveSentRoomFlag.LIVE:
+                    rooms_should_send.add(room_id)
```
Comment (MadLittleMods) [inline synapse/handlers/sliding_sync.py:633]: I understand why we always need to send a room if we `NEVER` sent it before (even if there aren't any updates) so that the client can get the base of meta info and `required_state`.

But I'm confused why we *always* have to send a room that was `PREVIOUSLY` sent down?

We should clarify this in the comments there.

File: synapse/types/handlers/__init__.py [hunk 11]
```diff
@@ -238,6 +238,14 @@ class StrippedHero:
         notification_count: int
         highlight_count: int
 
+        def __bool__(self) -> bool:
+            return (
+                self.initial
```
Comment (MadLittleMods) [inline synapse/types/handlers/__init__.py:243]:
```suggestion
                // If this is the first time the client is seeing the room, we should not filter it out
                // under any circumstance.
                self.initial
                // We need to let the client know if there are new events
```

File: synapse/handlers/sliding_sync.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (MadLittleMods) [reply synapse/handlers/sliding_sync.py]: Ohh, that `self.initial` being in the `def __bool__(self) -> bool:` condition is a tricky important detail here. Added a suggestion to add a comment there to point it out.

File: synapse/types/handlers/__init__.py [hunk 12]
```diff
@@ -367,7 +375,7 @@ def __bool__(self) -> bool:
         to tell if the notifier needs to wait for more events when polling for
         events.
         """
-        return bool(self.lists or self.rooms or self.extensions)
```
Comment (MadLittleMods) [inline synapse/types/handlers/__init__.py:370]: Just going over the possible scenario in my head where the sliding window `lists` could change order without any `rooms` updates but since we sort the rooms by `stream_ordering` of the latest activity, anything that would change the sort, would end up in `self.rooms` and cause us to send down the change.

Perhaps we should clarify that in a comment for why `self.lists` isn't included here though.

File: tests/rest/client/test_sync.py (hunk 3 above)
Comment (MadLittleMods) [reply tests/rest/client/test_sync.py:3571]: My inclination here was because this used the connection tracking, it was part of the extra layer that was changing.

But the nuance here is that everything is `initial: True` without it and this behavior is separate from connection tracking.

File: tests/rest/client/test_sync.py (hunk 4 above)
Comment (MadLittleMods) [reply tests/rest/client/test_sync.py]: :clap: Nice!! Happy to see that test up and running now :smiley:

File: tests/rest/client/test_sync.py [hunk 13]
```diff
@@ -4904,6 +4884,63 @@ def test_incremental_sync_full_state_never(self) -> None:
...
+                }
+            }
+        }
+
+        _, after_room_token = self.do_sync(sync_body, tok=user1_tok)
+
+        # Make the Sliding Sync request
+        response_body, _ = self.do_sync(
+            sync_body, since=after_room_token, tok=user1_tok
+        )
```
Comment (MadLittleMods) [inline tests/rest/client/test_sync.py:4913]: (align with other tests)

```suggestion
        _, from_token = self.do_sync(sync_body, tok=user1_tok)

        # Make the incremental Sliding Sync request
        response_body, _ = self.do_sync(
            sync_body, since=from_token, tok=user1_tok
        )
```

File: tests/rest/client/test_sync.py [hunk 14]
```diff
@@ -4904,6 +4884,63 @@ def test_incremental_sync_full_state_never(self) -> None:
...
+
+        # Nothing has happened in the room, so the room should not come down
+        # /sync.
+        self.assertIsNone(response_body["rooms"].get(room_id1))
+
+    def test_empty_room_comes_down_sync(self) -> None:
+        """
+        Test that rooms come down /sync even with empty required state and
+        timeline limit in initial sync.
+        """
```
Comment (MadLittleMods) [inline tests/rest/client/test_sync.py:4923]:
```suggestion
    def test_empty_initial_room_comes_down_sync(self) -> None:
        """
        Test that rooms come down /sync even with empty required state and
        timeline limit in initial sync.
        """
```

File: synapse/handlers/sliding_sync.py (hunk 9 above)
Comment (erikjohnston) [reply synapse/handlers/sliding_sync.py:638]: Yeah, I think constructing a new set is probably going to be more effort than this removes :shrug:

File: synapse/handlers/sliding_sync.py (hunk 10 above)
Comment (erikjohnston) [reply synapse/handlers/sliding_sync.py:638]: Have added a comment, but basically `PREVIOUSLY` indicates there was an update that we haven't sent down yet, so if the room is now within a list range (or a subscription) we should send it down with the missing update.

The other way to look at is the optimisation of "has anything happened in the rooms since `from_token`" only makes sense for rooms that are in the `LIVE` state, i.e. those are the rooms that we have sent down all updates for upto `from_token`.

**Review by MadLittleMods — APPROVED** [review]

File: synapse/handlers/sliding_sync.py (hunk 10 above)
Comment (erikjohnston) [reply synapse/handlers/sliding_sync.py]: Done. I've also added a clause to check (and ignore) `LIVE` so we can do an `assert_never(..)` (which means mypy will pick up if we add another flag and don't update this).

Also happy to just take what you suggested as well, though I think the slight ugliness is worth the the exhaustive matching. WDYT?

Comment (MadLittleMods) [reply synapse/handlers/sliding_sync.py]: Looks great! I agree that it's worth it. I've been wanting to do exhaustive matching stuff for the `membership` comparisons since I feel like there is too much wiggle room for things to slip through the cracks.
